Panic and reboot

## Bug Report

dhcp client runtime error?
I've catch it two time in one day. 

### Logs

```
panic: runtime error: slice bounds out of range [2:0]



goroutine 151398 [running]:

github.com/insomn[43206.598388] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000200

[43206.606496] CPU: 1 PID: 1627 Comm: init Tainted: G                T 5.10.69-talos #1

[43206.614361] Hardware name: Google Google Compute Engine/Google Compute Engine, BIOS Google 01/01/2011

[43206.623695] Call Trace:

iacslk/dhcp/dhcp[43206.626429]  dump_stack+0x6b/0x83

[43206.631184]  panic+0x106/0x2cd

v4/nclient4.udp.[43206.634470]  ? __hrtimer_init+0xc0/0xc0

[43206.639799]  do_exit.cold+0xaa/0xb6

[43206.643403]  do_group_exit+0x34/0xa0

destinationPort([43206.647146]  get_signal+0x15a/0x870

...)

	/.cache/m[43206.652157]  arch_do_signal+0x43/0x750

od/github.com/in[43206.657407]  exit_to_user_mode_prepare+0xb9/0x100

somniacslk/dhcp@[43206.663672]  syscall_exit_to_user_mode+0x28/0x150

v0.0.0-202108271[43206.669850]  entry_SYSCALL_64_after_hwframe+0x44/0xa9

73440-b95caade3e[43206.676432] RIP: 0033:0x4672e3

[43206.681002] Code: 24 20 c3 cc cc cc cc 48 8b 7c 24 08 8b 74 24 10 8b 54 24 14 4c 8b 54 24 18 4c 8b 44 24 20 44 8b 4c 24 28 b8 ca 00 00 00 0f 05 <89> 44 24 30 c3 cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc

[43206.699985] RSP: 002b:000000c000071e78 EFLAGS: 00000202 ORIG_RAX: 00000000000000ca

[43206.707843] RAX: fffffffffffffdfc RBX: 0000000000000000 RCX: 00000000004672e3

[43206.715094] RDX: 0000000000000000 RSI: 0000000000000080 RDI: 000000000475d618

[43206.722512] RBP: 000000c000071ec0 R08: 0000000000000000 R09: 0000000000000000

[43206.729774] R10: 000000c000071eb0 R11: 0000000000000202 R12: 000000c000071eb0

[43206.737036] R13: 000000c000d6a400 R14: 000000c0000004e0 R15: 00007f0ba7062ac3

ac/dhcpv4/nclien[43206.744731] Kernel Offset: 0x24c00000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)

[43206.756682] Rebooting in 10 seconds..

```


```
panic: runtime error: slice bounds out of range [2:0]



gorouti[43225.205229] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000200

[43225.213182] CPU: 1 PID: 2730 Comm: init Tainted: G                T 5.10.69-talos #1

[43225.221046] Hardware name: Google Google Compute Engine/Google Compute Engine, BIOS Google 01/01/2011

[43225.230384] Call Trace:

ne 151210 [runni[43225.233129]  dump_stack+0x6b/0x83

ng]:

github.com[43225.237925]  panic+0x106/0x2cd

/insomniacslk/dh[43225.242651]  do_exit.cold+0xaa/0xb6

[43225.247639]  do_group_exit+0x34/0xa0

cp/dhcpv4/nclien[43225.251380]  __x64_sys_exit_group+0xf/0x10

[43225.256949]  do_syscall_64+0x33/0x40

t4.udp.destinati[43225.260662]  entry_SYSCALL_64_after_hwframe+0x44/0xa9

onPort(...)

	/.[43225.267203] RIP: 0033:0x466c0b

[43225.271769] Code: db c5 ff ff cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc 8b 7c 24 08 b8 e7 00 00 00 0f 05 <c3> cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc cc 48

cache/mod/github[43225.290748] RSP: 002b:000000c000e59fa0 EFLAGS: 00000202 ORIG_RAX: 00000000000000e7

[43225.299999] RAX: ffffffffffffffda RBX: 000000c000e59fc0 RCX: 0000000000466c0b

[43225.307248] RDX: 0000000002c159a8 RSI: 0000000002c11858 RDI: 0000000000000002

[43225.314671] RBP: 000000c000e59fb0 R08: 0000000000000004 R09: 0000000000000000

[43225.321926] R10: 0000000001cee1f7 R11: 0000000000000202 R12: 000000c000e59c38

.com/insomniacsl[43225.329169] R13: 000000000009deb2 R14: 000000c00132eb60 R15: 0000000000000000

k/dhcp@v0.0.0-20[43225.338233] Kernel Offset: 0x1200000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)

[43225.350017] Rebooting in 10 seconds..

```
### Environment

- Talos version: v0.13.0-alpha.3-47-gd805ef290
- Kubernetes version: v1.22.2
- Platform: GCP

--- a/go.mod
+++ b/go.mod
@@ -36,6 +36,7 @@ require (
 	github.com/AlekSi/pointer v1.2.0
 	github.com/BurntSushi/toml v0.4.1
 	github.com/beevik/ntp v0.3.0
+	github.com/cenkalti/backoff/v4 v4.1.1
 	github.com/containerd/cgroups v1.0.2
 	github.com/containerd/containerd v1.5.7
 	github.com/containerd/cri v1.19.0
@@ -149,7 +150,6 @@ require (
 	github.com/bits-and-blooms/bitset v1.2.0 // indirect
 	github.com/blang/semver v3.5.1+incompatible // indirect
 	github.com/briandowns/spinner v1.6.1 // indirect
-	github.com/cenkalti/backoff/v4 v4.1.1 // indirect
 	github.com/cespare/xxhash/v2 v2.1.1 // indirect
 	github.com/chai2010/gettext-go v0.0.0-20160711120539-c6fed771bfd5 // indirect
 	github.com/cilium/ebpf v0.7.0 // indirect

--- a/internal/app/machined/pkg/controllers/network/operator_spec.go
+++ b/internal/app/machined/pkg/controllers/network/operator_spec.go
@@ -8,7 +8,9 @@ import (
 	"context"
 	"fmt"
 	"sync"
+	"time"
 
+	"github.com/cenkalti/backoff/v4"
 	"github.com/cosi-project/runtime/pkg/controller"
 	"github.com/cosi-project/runtime/pkg/resource"
 	"github.com/cosi-project/runtime/pkg/state"
@@ -91,7 +93,7 @@ type operatorRunState struct {
 	wg     sync.WaitGroup
 }
 
-func (state *operatorRunState) Start(ctx context.Context, notifyCh chan<- struct{}) {
+func (state *operatorRunState) Start(ctx context.Context, notifyCh chan<- struct{}, logger *zap.Logger, id string) {
 	state.wg.Add(1)
 
 	ctx, state.cancel = context.WithCancel(ctx)
@@ -99,8 +101,46 @@ func (state *operatorRunState) Start(ctx context.Context, notifyCh chan<- struct
 	go func() {
 		defer state.wg.Done()
 
-		state.Operator.Run(ctx, notifyCh)
+		state.runWithRestarts(ctx, notifyCh, logger, id)
 	}()
+}
+
+func (state *operatorRunState) runWithRestarts(ctx context.Context, notifyCh chan<- struct{}, logger *zap.Logger, id string) {
+	backoff := backoff.NewExponentialBackOff()
+
+	// disable number of retries limit
+	backoff.MaxElapsedTime = 0
+
+	for ctx.Err() == nil {
+		if err := state.runWithPanicHandler(ctx, notifyCh, logger, id); err == nil {
+			// operator finished without an error
+			return
+		}
+
+		interval := backoff.NextBackOff()
+
+		logger.Debug("restarting operator", zap.Duration("interval", interval), zap.String("operator", id))
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(interval):
+		}
+	}
+}
+
+func (state *operatorRunState) runWithPanicHandler(ctx context.Context, notifyCh chan<- struct{}, logger *zap.Logger, id string) (err error) {
+	defer func() {
+		if p := recover(); p != nil {
+			err = fmt.Errorf("panic: %v", p)
+
+			logger.Error("operator panicked", zap.Stack("stack"), zap.Error(err), zap.String("operator", id))
+		}
+	}()
+
+	state.Operator.Run(ctx, notifyCh)
+
+	return nil
 }
 
 func (state *operatorRunState) Stop() {
@@ -210,7 +250,7 @@ func (ctrl *OperatorSpecController) reconcileOperators(ctx context.Context, r co
 			}
 
 			logger.Debug("starting operator", zap.String("operator", id))
-			ctrl.operators[id].Start(ctx, notifyCh)
+			ctrl.operators[id].Start(ctx, notifyCh, logger, id)
 		}
 	}
 

--- a/internal/app/machined/pkg/controllers/network/operator_spec_test.go
+++ b/internal/app/machined/pkg/controllers/network/operator_spec_test.go
@@ -44,6 +44,7 @@ type OperatorSpecSuite struct {
 type mockOperator struct {
 	spec     network.OperatorSpecSpec
 	notifyCh chan<- struct{}
+	panicked bool
 
 	mu          sync.Mutex
 	addresses   []network.AddressSpecSpec
@@ -72,13 +73,22 @@ func (mock *mockOperator) Run(ctx context.Context, notifyCh chan<- struct{}) {
 		runningOperatorsMu.Unlock()
 	}
 
-	<-ctx.Done()
-
-	{
+	defer func() {
 		runningOperatorsMu.Lock()
 		delete(runningOperators, mock.Prefix())
 		runningOperatorsMu.Unlock()
+	}()
+
+	if mock.spec.Operator == network.OperatorDHCP6 {
+		// DHCP6 operator panics on odd run
+		if !mock.panicked {
+			mock.panicked = true
+
+			panic("oh no, IPv6!!!")
+		}
 	}
+
+	<-ctx.Done()
 }
 
 func (mock *mockOperator) notify() {
@@ -330,6 +340,48 @@ func (suite *OperatorSpecSuite) TestScheduling() {
 	suite.Assert().NoError(retry.Constant(3*time.Second, retry.WithUnits(100*time.Millisecond)).Retry(
 		func() error {
 			return suite.assertRunning([]string{"vip/eth0"}, func(op *mockOperator) error {
+				return nil
+			})
+		}))
+}
+
+func (suite *OperatorSpecSuite) TestPanic() {
+	specPanic := network.NewOperatorSpec(network.NamespaceName, "dhcp6/eth0")
+	*specPanic.TypedSpec() = network.OperatorSpecSpec{
+		Operator:  network.OperatorDHCP6,
+		LinkName:  "eth0",
+		RequireUp: true,
+		DHCP6: network.DHCP6OperatorSpec{
+			RouteMetric: 1024,
+		},
+	}
+
+	suite.Require().NoError(suite.state.Create(suite.ctx, specPanic))
+
+	linkState := network.NewLinkStatus(network.NamespaceName, "eth0")
+	*linkState.TypedSpec() = network.LinkStatusSpec{
+		OperationalState: nethelpers.OperStateUp,
+	}
+
+	suite.Require().NoError(suite.state.Create(suite.ctx, linkState))
+
+	// DHCP6 operator should panic and then restart
+	suite.Assert().NoError(retry.Constant(3*time.Second, retry.WithUnits(100*time.Millisecond)).Retry(
+		func() error {
+			return suite.assertRunning([]string{"dhcp6/eth0"}, func(op *mockOperator) error { return nil })
+		}))
+
+	// bring down the interface, operator should be stopped
+	_, err := suite.state.UpdateWithConflicts(suite.ctx, linkState.Metadata(), func(r resource.Resource) error {
+		r.(*network.LinkStatus).TypedSpec().OperationalState = nethelpers.OperStateDown
+
+		return nil
+	})
+	suite.Require().NoError(err)
+
+	suite.Assert().NoError(retry.Constant(3*time.Second, retry.WithUnits(100*time.Millisecond)).Retry(
+		func() error {
+			return suite.assertRunning(nil, func(op *mockOperator) error {
 				return nil
 			})
 		}))

